### PR TITLE
Optimized vec::IntoIter::next_chunk impl

### PR DESCRIPTION
**Pull request by the8472**
```
x86_64v1, default
test vec::bench_next_chunk                               ... bench:         696 ns/iter (+/- 22)
x86_64v1, pr
test vec::bench_next_chunk                               ... bench:         309 ns/iter (+/- 4)

znver2, default
test vec::bench_next_chunk                               ... bench:      17,272 ns/iter (+/- 117)
znver2, pr
test vec::bench_next_chunk                               ... bench:         211 ns/iter (+/- 3)
```

On znver2 the default impl seems to be slow due to different inlining decisions. It goes through `core::array::iter_next_chunk`
which has a deep call tree.

### CODE REVIEW

File: library/alloc/src/vec/into_iter.rs [hunk 1]
```diff
@@ -204,6 +204,43 @@ impl<T, A: Allocator> Iterator for IntoIter<T, A> {
...
+                ptr::copy_nonoverlapping(self.ptr, raw_ary.as_mut_ptr() as *mut T, len);
+                self.forget_remaining_elements();
+                return Err(array::IntoIter::new_unchecked(raw_ary, 0..len));
+            }
+        }
+
+        // Safety: `len` is larger than the array size. Copy a fixed amount here to fully initialize
+        // the array.
+        return unsafe {
+            ptr::copy_nonoverlapping(self.ptr as *const T, raw_ary.as_mut_ptr() as *mut T, N);
```
Comment (Mark-Simulacrum) [inline library/alloc/src/vec/into_iter.rs:238]: The cast to `*const T` here seems like it should not be necessary?

**Comment by Mark-Simulacrum** [comment]
r=me

@bors rollup=never in case has unexpected perf effects

File: library/alloc/src/vec/into_iter.rs (hunk 1 above)
Comment (the8472) [reply library/alloc/src/vec/into_iter.rs]: fixed and added a test.

**Pull request merged**